zrml/shares/src/lib.rs

```diff
@@ -1,313 +1,143 @@
-            let to = T::Lookup::lookup(dest)?;
-            <Self as Shares<T::AccountId, T::Hash>>::transfer(share_id, &from, &to, amount)?;
-
-            Self::deposit_event(RawEvent::Transferred(share_id, from, to, amount));
-        }
-
-        /// Wraps the native currency into a "share" so that it can be used as if it was part of this
-        /// pallet.
-        #[weight = 0]
-        pub fn wrap_native_currency(origin, amount: BalanceOf<T>) {
-            let sender = ensure_signed(origin)?;
-
-            Self::do_wrap_native_currency(sender, amount)?;
-        }
-
-        #[weight = 0]
-        pub fn unwrap_native_currency(origin, amount: BalanceOf<T>) {
-            let sender = ensure_signed(origin)?;
-
-            Self::do_unwrap_native_currency(sender, amount)?;
-        }
-    }
-}
-
-impl<T: Trait> Module<T> {
-    pub fn set_balance(
-        share_id: T::Hash,
-        who: &T::AccountId,
-        balance: BalanceOf<T>,
-    ) -> DispatchResult {
-        <Accounts<T>>::mutate(share_id, who, |data| data.free = balance);
-        Ok(())
-    }
-
-    pub fn set_reserved(
-        share_id: T::Hash,
-        who: &T::AccountId,
-        reserved: BalanceOf<T>,
-    ) -> DispatchResult {
-        <Accounts<T>>::mutate(share_id, who, |data| data.reserved = reserved);
-        Ok(())
-    }
-
-    #[inline]
-    fn get_module_id() -> T::AccountId {
-        T::ModuleId::get().into_account()
     }
 }
 
-impl<T: Trait> Shares<T::AccountId, T::Hash> for Module<T> {
+impl<T> Shares<T::AccountId> for Module<T>
+where
+    T: Trait,
+{
     type Balance = BalanceOf<T>;
+    type Hash = T::Hash;
 
-    fn free_balance(share_id: T::Hash, who: &T::AccountId) -> BalanceOf<T> {
-        Self::accounts(share_id, who).free
+    fn free_balance(share_id: Asset<Self::Hash>, who: &T::AccountId) -> BalanceOf<T> {
+        T::Currency::free_balance(share_id, who)
     }
 
-    fn total_supply(share_id: T::Hash) -> BalanceOf<T> {
-        <TotalSupply<T>>::get(share_id)
+    fn total_supply(share_id: Asset<Self::Hash>) -> BalanceOf<T> {
+        T::Currency::total_issuance(share_id)
     }
 
-    fn destroy(share_id: T::Hash, from: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
-        if amount.is_zero() {
-            return Ok(());
-        }
-
-        Self::ensure_can_withdraw(share_id, from, amount)?;
-
-        <TotalSupply<T>>::mutate(share_id, |am| *am -= amount);
-        Self::set_balance(share_id, from, Self::free_balance(share_id, from) - amount)?;
-
+    fn destroy(
+        share_id: Asset<Self::Hash>,
+        from: &T::AccountId,
+        amount: BalanceOf<T>,
+    ) -> DispatchResult {
+        T::Currency::slash(share_id, from, amount);
         Ok(())
     }
 
-    fn destroy_all(share_id: T::Hash) -> DispatchResult {
-        <Accounts<T>>::remove_prefix(share_id);
-        <TotalSupply<T>>::remove(share_id);
-
-        Ok(())
+    fn destroy_all(_share_id: Asset<Self::Hash>) -> DispatchResult {
+        todo!()
```

Comment:
Not sure if this is possible because to implement.

Reply:
Yeah doesn't seem possible without modifying orml_tokens to add it

Reply:
Everything else is good though. Perhaps 'shares' is no longer needed

Reply:
Oh wait, I think we could do something like what's done in frame_system [here](https://github.com/paritytech/substrate/blob/master/frame/system/src/lib.rs#L419) By calling [`kill_prefix`](https://github.com/paritytech/substrate/blob/master/frame/support/src/storage/unhashed.rs#L95) directly on the prefix of the storage

Reply:
NVM, the double map for [account storage in orml_tokens](https://github.com/open-web3-stack/open-runtime-module-library/blob/0.2.0/tokens/src/lib.rs#L188) is keyed first by `AccountId` which means there will not be a shared prefix for an entire shares type, like before. Bummer.